### PR TITLE
Add 'exact' keyword in span_range, span and interval methods 

### PR DESCRIPTION
**Pull request by svanita00**
## Pull Request Checklist

Thank you for taking the time to improve Arrow! Before submitting your pull request, please check all *appropriate* boxes:

- [x] 🧪  Added **tests** for changed code.
- [x] 🛠️  All tests **pass** when run locally (run `tox` or `make test` to find out!).
- [x] 🧹  All linting checks **pass** when run locally (run `tox -e lint` or `make lint` to find out!).
- [ ] 📚  Updated **documentation** for changed code.
- [x] ⏩  Code is **up-to-date** with the `master` branch.

If you have *any* questions about your code changes or any of the points above, please submit your questions along with the pull request and we will try our best to help!

## Description of Changes

Merged PR #817 with updated master. Added edge cases in span_range function when exact keyword is True. Added correlated tests for span_range. 

Closes: #498

### CODE REVIEW

**Comment by svanita00** [comment]
I'm not sure why the lint is failing here, I ran it locally and it passed there..

File: arrow/arrow.py [hunk 1]
```diff
@@ -387,7 +387,7 @@ def span(self, frame, count=1, bounds="[)"):
             whether to include or exclude the start and end values in the span. '(' excludes
             the start, '[' includes the start, ')' excludes the end, and ']' includes the end.
             If the bounds are not specified, the default bound '[)' is used.
-
+        :param exact: (optional) whether to return a timespan that starts exacly at specificed start
```
Comment (systemcatch) [inline arrow/arrow.py:390]: The reason for the lint (docs really) command failing is because there needs to be a blank line between the params and the next part of the docstring.

When I add that in everything passes fine, try running `tox -e docs` or `make docs` on your machine to see if it passes.

Comment (systemcatch) [reply arrow/arrow.py:390]: There are also some typos in this line, why don't you use something similar to the `span_range` docstring?

**Review by systemcatch — CHANGES_REQUESTED** [review]
Hey @svanita00 I've left some initial comments.

**Comment by systemcatch** [comment]
Seems like the bug found in the old PR is now fixed. :+1: 

```python
from datetime import datetime
import arrow

start = datetime(2013, 5, 5, 12, 30)
end = datetime(2013, 5, 5, 12, 40)

for r in arrow.Arrow.span_range('minute', start, end, exact=True):
    print(r)
```

Old bug
```shell
(<Arrow [2013-05-05T12:30:00+00:00]>, <Arrow [2013-05-05T12:30:59.999999+00:00]>)
(<Arrow [2013-05-05T12:31:00+00:00]>, <Arrow [2013-05-05T12:31:59.999999+00:00]>)
(<Arrow [2013-05-05T12:32:00+00:00]>, <Arrow [2013-05-05T12:32:59.999999+00:00]>)
(<Arrow [2013-05-05T12:33:00+00:00]>, <Arrow [2013-05-05T12:33:59.999999+00:00]>)
(<Arrow [2013-05-05T12:34:00+00:00]>, <Arrow [2013-05-05T12:34:59.999999+00:00]>)
(<Arrow [2013-05-05T12:35:00+00:00]>, <Arrow [2013-05-05T12:35:59.999999+00:00]>)
(<Arrow [2013-05-05T12:36:00+00:00]>, <Arrow [2013-05-05T12:36:59.999999+00:00]>)
(<Arrow [2013-05-05T12:37:00+00:00]>, <Arrow [2013-05-05T12:37:59.999999+00:00]>)
(<Arrow [2013-05-05T12:38:00+00:00]>, <Arrow [2013-05-05T12:38:59.999999+00:00]>)
(<Arrow [2013-05-05T12:39:00+00:00]>, <Arrow [2013-05-05T12:39:59.999999+00:00]>)
(<Arrow [2013-05-05T12:40:00+00:00]>, <Arrow [2013-05-05T12:39:59.999999+00:00]>)
```

New correct
```shell
(<Arrow [2013-05-05T12:30:00+00:00]>, <Arrow [2013-05-05T12:30:59.999999+00:00]>)
(<Arrow [2013-05-05T12:31:00+00:00]>, <Arrow [2013-05-05T12:31:59.999999+00:00]>)
(<Arrow [2013-05-05T12:32:00+00:00]>, <Arrow [2013-05-05T12:32:59.999999+00:00]>)
(<Arrow [2013-05-05T12:33:00+00:00]>, <Arrow [2013-05-05T12:33:59.999999+00:00]>)
(<Arrow [2013-05-05T12:34:00+00:00]>, <Arrow [2013-05-05T12:34:59.999999+00:00]>)
(<Arrow [2013-05-05T12:35:00+00:00]>, <Arrow [2013-05-05T12:35:59.999999+00:00]>)
(<Arrow [2013-05-05T12:36:00+00:00]>, <Arrow [2013-05-05T12:36:59.999999+00:00]>)
(<Arrow [2013-05-05T12:37:00+00:00]>, <Arrow [2013-05-05T12:37:59.999999+00:00]>)
(<Arrow [2013-05-05T12:38:00+00:00]>, <Arrow [2013-05-05T12:38:59.999999+00:00]>)
(<Arrow [2013-05-05T12:39:00+00:00]>, <Arrow [2013-05-05T12:39:59.999999+00:00]>)
```

File: arrow/arrow.py (hunk 1 above)
Comment (svanita00) [reply arrow/arrow.py]: Got it! I fixed these in the latest commit.

File: arrow/arrow.py [hunk 2]
```diff
@@ -528,12 +538,26 @@ def span_range(cls, frame, start, end, tz=None, limit=None, bounds="[)"):
...
-        return (r.span(frame, bounds=bounds) for r in _range)
+        for r in _range:
+            if exact:
+                floor, ceil = r.span(frame, bounds=bounds, exact=exact)
+                if ceil > end:
+                    ceil = end
+                    if bounds[1] == ")":
+                        ceil += relativedelta(microseconds=-1)
+                if floor == end:
+                    break  # TODO: implementation detail: break OR yield floor, floor IF bounds[1] == "]"
```
Comment (systemcatch) [inline arrow/arrow.py:552]: We should `break` here.

File: tests/test_arrow.py [hunk 3]
```diff
@@ -1623,6 +1791,16 @@ def test_bounds_are_validated(self):
...
+    def test_exact(self):
+
+        result_floor, result_ceil = self.arrow.span("hour", exact=True)
+
+        expected_floor = datetime(2013, 2, 15, 3, 41, 22, 8923, tzinfo=tz.tzutc())
+        expected_ceil = datetime(2013, 2, 15, 4, 41, 22, 8922, tzinfo=tz.tzutc())
+
+        assert result_floor == expected_floor
+        assert result_ceil == expected_ceil
+
```
[the remark this answers is not in the document]
Comment (svanita00) [reply tests/test_arrow.py:1803]: Yes, that makes sense. I just added a few tests with different bounds and a test using exact, count, and bounds in the parameters in the last commit

**Review by systemcatch — APPROVED** [review]
Thanks @svanita00 we're all good to go here. 👍
